OMS Troubleshooting Tool

aswatt opened the pull request:
Apologies this took so long to create!

This pull request includes all of the code included in the troubleshooting tool. Currently all of the code is separate from the agent itself - will include a separate pull request later focusing on integrating the tool into the codebase to allow it to install when the agent is installed. Thank you!

hestolz commented on source/code/troubleshooter/omsagent_tsg_zip/tsg_files/tsg_code/syslog1/tsg_checkconf.py:
```diff
@@ -0,0 +1,161 @@
+import os
```
how come the directory is `syslog1` not just `syslog`?

aswatt replied:
I ran into some errors when I named it `syslog `(mainly I believe there's a python package or something called syslog), which went away as soon as I renamed the folder, I just called it `syslog1` as a placeholder because I couldn't think of a better name for it

hestolz commented on source/code/troubleshooter/omsagent_tsg_zip/tsg_files/tsg_code/tsg_general.py:
```diff
@@ -0,0 +1,191 @@
+        "  - Workspace Name\n"\
+        "  - Region Workspace is located\n"\
+        "  - Pricing Tier assigned to the Workspace\n"\
+        "(The above points can all be found in Azure Support Center.)\n"\
```
Not sure what this means to the customer. I thought ASC is an internal tool just for us and CSS? Or is this also the name for the portal where customers go to file a ticket?

aswatt replied:
Apologies you're right, I originally took this from some documentation meant for CSS but forgot to change that, changed below:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/source/code/troubleshooter/modules/main.py#L240-L249

hestolz commented on source/code/troubleshooter/omsagent_tsg_zip/tsg_files/tsg_code/tsg_errors.py:
```diff
@@ -0,0 +1,323 @@
+warnings = {115, 127, 151}
+
+# dictionary correlating error codes to error messages
+tsg_error_codes = {
```
like I mentioned above, IMO ideally these would be given informative variable names, otherwise it is a bit hard to follow the code without having to reference back here frequently

aswatt replied:
Yeah, I changed these as well to have names through the same `tsg_error_codes.py` file:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/source/code/troubleshooter/modules/error_codes.py#L8-L12
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/source/code/troubleshooter/modules/errors.py#L21

hestolz commented on source/code/troubleshooter/omsagent_tsg_zip/tsg_files/tsg_code/tsg_errors.py:
```diff
@@ -0,0 +1,323 @@
+    154 : "This version of Python is missing the {0} package. (You can check by opening up "\
+          "python and typing 'import {0}'.) Please install this package and run the troubleshooter again."
+
+}  # TODO: keep up to date with error codes onenote
```
Might be better to maintain an .md in this repo publicly so that all users can see?

aswatt replied:
Added it to the TST Github documentation here:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/docs/Troubleshooting-Tool.md#list-of-possible-errors

hestolz commented on source/code/troubleshooter/omsagent_tsg_zip/tsg_files/tsg_code/tsg_general.py:
```diff
@@ -0,0 +1,191 @@
+
+    # create command to run
+    logcollect_cmd = ['sudo', 'sh', logcollect_sh_path]
+    if (sr_num != ''):
```
isn't SR number mandatory?

aswatt replied:
I didn't realize it was mandatory, sorry! Updated the log collector to make it mandatory:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/source/code/troubleshooter/modules/main.py#L116

hestolz commented on source/code/troubleshooter/omsagent_tsg_zip/tsg_files/tsg_code/syslog1/tsg_checkconf.py:
```diff
@@ -0,0 +1,161 @@
+from tsg_info             import tsginfo_lookup
+from install.tsg_checkoms import comp_versions_ge, get_oms_version
+
+syslogconf_path = "/etc/opt/microsoft/omsagent/sysconf/omsagent.d/syslog.conf"
```
why is this in `/sysconf/` and not the `workspaceid/conf` directory?

aswatt replied:
Mistake on my part, updated the correct path:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/source/code/troubleshooter/modules/syslog1/check_conf.py#L10

hestolz commented on source/code/troubleshooter/omsagent_tsg_zip/tsg_files/tsg_code/connect/tsg_checke2e.py:
```diff
@@ -0,0 +1,71 @@
+    failures = []
+
+    print("--------------------------------------------------------------------------------")
+    print(" Please go to https://ms.portal.azure.com and navigate to your workspace.\n"\
```
I believe for it is just `portal.azure.com`, `ms.` is just for internal Microsoft

aswatt replied:
Didn't realize that, updated the link accordingly:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/source/code/troubleshooter/modules/connect/check_e2e.py#L18

hestolz commented on source/code/troubleshooter/omsagent_tsg_zip/tsg_files/tsg_code/connect/tsg_checkendpts.py:
```diff
@@ -0,0 +1,82 @@
+
+omsadmin_path = "/etc/opt/microsoft/omsagent/conf/omsadmin.conf"
+
+# ping specific endpoint
```
any chance to add something above a ping test? e.g. a `curl` call or some other way to test for actual connection success. i think in many cases ping will succeed but that just says that the destination is reachable; if an actual connection is attempted it may be denied by any firewall rules the customer has applied.

aswatt replied:
Sorry the comment here was misleading, it's actually a socket test, updated the comment here:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/source/code/troubleshooter/modules/connect/check_endpts.py#L11
But I also have plans to improve this test by also using SSL checks and the certs, but that's for a future release

hestolz commented on source/code/troubleshooter/omsagent_tsg_zip/tsg_files/tsg_code/custom_logs/tsg_checkclconf.py:
```diff
@@ -0,0 +1,106 @@
+
+    # <15 minutes since last applying config
+    else:
+        print(" Please try waiting for the agent to pull the config from the OMS backend.")
```
suggestion, include the command they can use to manually pull 
`sudo su omsagent -c 'python /opt/microsoft/omsconfig/Scripts/PerformRequiredConfigurationChecks.py'`

aswatt replied:
Updated the print statement to include the command:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/source/code/troubleshooter/modules/custom_logs/check_clconf.py#L16-L20

hestolz commented on source/code/troubleshooter/omsagent_tsg_zip/tsg_files/tsg_code/high_cpu_mem/tsg_high_cpu_mem.py:
```diff
@@ -0,0 +1,74 @@
+        success = print_errors(checked_highcpu)
+
+    # check slab memory / dentry cache issue
+    print("Checking slab memory / dentry cache usage...")
```
so are we only checking for these specific types of mem usage, not overall?

aswatt replied:
At the moment the focus is on those types, but I plan to add more checks in the future regarding all memory usage

aswatt commented on source/code/troubleshooter/omsagent_tsg_zip/tsg_files/tsg_code/heartbeat/tsg_heartbeat.py:
```diff
@@ -0,0 +1,211 @@
+        err_info = err_matches.groups()[0]
+        # check if permission error
+        if (err_info == "This script must be run as root or as the omsagent user."):
+            tsg_error_info.append((omsadmin_sh_path,))
```
Actually for this error it only needs `omsadmin_sh_path`, but it needs to be entered as a tuple, thus the weird formatting of `(<item>,)`

aswatt commented:
(Apologies about the weird git history, I don't think Github liked the fact that I changed the folders where all the code was stored)
Sorry this took so long! Main points to this update are:
- added troubleshooter install to beginning of shell bundle build (would appreciate an extra set of eyes on this, commit 741f655242248e56b6b07fde31dff2fc449fe53d)
- added the ability to run in silent mode vs. interactive mode
- cleaned up file organization
- added variable names to all error codes, for readability
- fixed compatibility with Python 2.6
- various bug fixes
Plus various other additions. Thank you so much for your patience!

aswatt commented on installer/bundle/bundle_skel.sh:
```diff
@@ -770,6 +843,8 @@ install_extra_package()
 ulinux_detect_installer
 set -e
```
Move to after install_troubleshooter

aswatt replied:
Fixed here:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/installer/bundle/bundle_skel.sh#L846-L849

aswatt commented on installer/bundle/bundle_skel.sh:
```diff
@@ -387,6 +411,55 @@ install_if_program_does_not_exist_on_system()
+    # grab tst bundle
+    wget $TST_PKG > /dev/null 2>&1
+    if [ $? -ne 0 ]; then
+        echo "Error downloading troubleshooter. To install it manually, please go to the below link:"
```
Add in comment about how it will try to install with OMS install

aswatt replied:
Fixed in these places:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/installer/bundle/bundle_skel.sh#L425
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/installer/bundle/bundle_skel.sh#L433
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/installer/bundle/bundle_skel.sh#L456

simathih commented on installer/bundle/bundle_skel.sh:
```diff
@@ -114,6 +121,23 @@ source_references()
 EOF
 }
 
```
please ensure cleanup is done during uninstall

aswatt replied:
Purge script appears to already remove `/opt/microsoft/omsagent` folder (which includes all of the code for the troubleshooter)

Location in uninstall.sh script:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/9e4c1efb9eaeeafd44d5b9e05c75828534976bd0/installer/scripts/uninstall.sh#L150

Location in bundle_skel.sh script (called if uninstall.sh script isn't found):
https://github.com/microsoft/OMS-Agent-for-Linux/blob/9e4c1efb9eaeeafd44d5b9e05c75828534976bd0/installer/bundle/bundle_skel.sh#L1143

aswatt commented on source/code/troubleshooter/modules/errors.py:
```diff
@@ -0,0 +1,402 @@
+    err_codes = {0 : "No errors found"}
+    section_name = "{0} Error Codes".format(err_type)
+    try:
+        ts_doc = urlopen(TSG_URL)
```
Change error of can't connect to Github to warning (bc not necessary for rest of troubleshooter)

aswatt replied:
Fixed here:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/source/code/troubleshooter/modules/errors_tsg.py#L58-L64

aswatt commented on source/code/troubleshooter/modules/install/check_oms.py:
```diff
@@ -0,0 +1,98 @@
+        return ERR_OLD_OMS_VER
+
+    # get most recent version
+    updated_curr_oms_version = update_curr_oms_version()
```
Change from error to warning if can't reach Github

aswatt replied:
Fixed here:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/source/code/troubleshooter/modules/install/check_oms.py#L98-L102

aswatt commented on source/code/troubleshooter/modules/syslog1/check_conf.py:
```diff
@@ -0,0 +1,171 @@
+
+OMSADMIN_PATH = "/etc/opt/microsoft/omsagent/conf/omsadmin.conf"
+SYSLOGCONF_PATH = "/etc/opt/microsoft/omsagent/conf/omsagent.d/syslog.conf"
+SYSLOGDEST_PATH = ""
```
Add path in here

aswatt commented on source/code/troubleshooter/omsagent_tsg_zip/tsg_files/omsagent_tsg:
```diff
@@ -0,0 +1,91 @@
+{
+    # check python
+    echo "Checking for python install..."
+    which python 1> /dev/null 2> /dev/null
```
Added this in here:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/source/code/troubleshooter/troubleshooter#L13-L14
and here:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/source/code/troubleshooter/troubleshooter#L29-L30

aswatt commented on source/code/troubleshooter/omsagent_tsg_zip/install_tsg:
```diff
@@ -0,0 +1,39 @@
+
+# copy over files
+echo "Installing troubleshooter on machine..."
+mkdir $plugin_tsg_path
```
Changed all instances of `mkdir` to `mkdir -p` just in case:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/source/code/troubleshooter/install_tst#L32

aswatt commented on source/code/troubleshooter/omsagent_tsg_zip/tsg_files/tsg_code/heartbeat/tsg_heartbeat.py:
```diff
@@ -0,0 +1,211 @@
+        return 0
+    elif (is_running == 0):
+        # don't have any extra info
+        return 122
```
Added a file `error_codes.py` which correlates all the magic numbers to more descriptive names (which are then used throughout):
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/source/code/troubleshooter/modules/error_codes.py#L1

hestolz replied:
beautiful

aswatt commented on source/code/troubleshooter/omsagent_tsg_zip/tsg_files/tsg_code/heartbeat/tsg_heartbeat.py:
```diff
@@ -0,0 +1,211 @@
+        return 200  # TODO: fix error code
+
+def check_omsagent_running_omsadmin(workspace):
+    output = subprocess.check_output(['sh', omsadmin_sh_path, '-l'], universal_newlines=True)
```
Monkey patched it in:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/source/code/troubleshooter/modules/main.py#L15-L35

aswatt commented on source/code/troubleshooter/omsagent_tsg_zip/tsg_files/tsg_tools/datafiles/linux.data:
```diff
@@ -0,0 +1,187 @@
+%Variables
```
Apologies that was my plan originally (but forgot that I left in the files for testing). Added the files to `base_omsagent.data` to be copied over at install:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/installer/datafiles/base_omsagent.data#L145-L149

aswatt commented on source/code/troubleshooter/omsagent_tsg_zip/tsg_files/tsg_tools/Troubleshooting.md:
```diff
@@ -0,0 +1,528 @@
+# Troubleshooting Guide for OMS Agent for Linux
```
Grabbing the TSG via URL when needed:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/bff7c80f869d6f5a4046ed41b4bd0ed89d69219f/source/code/troubleshooter/modules/errors_tsg.py#L11

hestolz commented on installer/bundle/bundle_skel.sh:
```diff
@@ -768,6 +844,8 @@ install_extra_package()
 #
 
 ulinux_detect_installer
+install_troubleshooter
```
this is still not the best place for this to be, notice if customer is running shell bundle with `--purge` or w/e, you will first try installing the troubleshooter and then it will be removed again. I think the ideal place for this would be only in the execution path for `--install` or `--upgrade`, at the END of the install/upgrade process. That way there is the least risk for interference. You may want to `set +e` before calling and then reset with `set -e` afterwards, so that the script doesn't exit if there is some uncaught exception. Just make sure that, even if it is the last thing that executes, a failure `install_troubleshooter` won't cause the bundle exit code to be nonzero

aswatt replied:
That makes sense, one of my main thoughts though was that I wanted it to try and install at the beginning of the install process, in case OMS install failed the troubleshooter would still be on the machine

aswatt replied:
But as long as it'll install at the end even if the shell bundle exits out then that works better even

hestolz replied:
in that case maybe put at beginning of install or upgrade process
just make sure to `set +e` or w/e method to prevent total failure on tst fail

aswatt replied:
Looking at Dmitry's comment below, adding the troubleshooter install to the end of the install process does seem like a better option (where it will only install via wget if OMS install is set to return a non-0 exit code, since that means the troubleshooter couldn't install through the base_omsagent.data files), so I think you're right in putting it at the end (with the `set +e` and `set -e` bookends around it of course)

aswatt replied:
I updated the installation to occur at the end, and only when `--install` or `--upgrade` are used (see https://github.com/microsoft/OMS-Agent-for-Linux/pull/1190#issuecomment-663785862 for more in-depth analysis), hopefully it shouldn't interfere with OMS installation at all

Dmitry-Matveev commented on installer/bundle/bundle_skel.sh:
```diff
@@ -387,6 +411,58 @@ install_if_program_does_not_exist_on_system()
+    cd $TST_EXTRACT_DIR
+
+    # grab tst bundle
+    wget $TST_PKG > /dev/null 2>&1
```
Is "wget" a backup option in case we could not install from the setup files or is it a main installation option?
I remember it was discussed briefly during the initial review, but I was not sure which mode is the main one.
Overall, starting with network may not be a good option - in a fairly significant number of cases certain URLs would be blocked by customer VM firewall and we won't be able to install t-shoot tool to find that firewall is an issue :)

This comment may apply to the places in t-shoot tool code that rely on certain URLs being available (such as docs/KB/TSGs download). We can move such code behind connection checks where possible.

Dmitry-Matveev replied:
Also, the tool would, probably, need to disclose all external locations we use to download pieces we need. E.g. in the initial dialog it presents or before certain t-shoot option is executed. You can follow up with Luke on this one to see if this is still a requirement. It used to be couple years ago when we delivered t-shoot tool for Azure Monitor's Status Monitor.

aswatt replied:
That's valid, I originally had it that the 'wget' was the main installation (in case the installation of the troubleshooter failed), but it might actually be better if I have the main way of installing it be normally through the files, and only try to use the wget link OMS installation fails in some way

aswatt replied:
I updated it so `wget` is used as a backup only if the troubleshooter isn't on the machine (see https://github.com/microsoft/OMS-Agent-for-Linux/pull/1190#issuecomment-663785862 for more detail), and if the `wget` doesn't work I have it print out the link needed to manually install the troubleshooter (aka the same link that didn't work), which should cover installation documentation

aswatt replied:
I believe that inside the troubleshooter, all of the external links that I reference for information (i.e. not for connection checks) are all from that one Github repo (and the error message prints out the link that couldn't be reached on failure), but do you think I should add some more information at the start of the troubleshooter or something along those lines?

Dmitry-Matveev commented on installer/bundle/bundle_skel.sh:
```diff
@@ -387,6 +411,58 @@ install_if_program_does_not_exist_on_system()
     fi
 }
 
+install_troubleshooter()
```
I'd second Henry's point below, t-shoot tool installation should not add a failure mode for OMS agent installation or cleanup/purge. Can you please d-check or add a setup test (if we have such infra at all at the moment) that failure in t-shoot setup does not prevent OMS setup and that having t-shoot tool installed does not prevent OMS uninstall or leaves leftovers?

aswatt replied:
I changed it so `install_troubleshooter` isn't called until the end of OMS install (in the `cleanup_and_exit` script, so as to limit the possible interference), as well as some other assurances that it won't affect OMS installation / cleanup even if it fails (see more details at https://github.com/microsoft/OMS-Agent-for-Linux/pull/1190#issuecomment-663785862 )

Dmitry-Matveev commented on installer/datafiles/base_omsagent.data:
```diff
@@ -142,6 +142,59 @@ MAINTAINER:              'Microsoft Corporation'
+/opt/microsoft/omsagent/tst/modules/high_cpu_mem/omiHighCPUDiagnosticsTST.sh;   source/code/troubleshooter/modules/high_cpu_mem/omiHighCPUDiagnosticsTST.sh;    644; root; root
+/opt/microsoft/omsagent/tst/modules/high_cpu_mem/__init__.py;                   source/code/troubleshooter/modules/high_cpu_mem/__init__.py;                    644; root; root
+
+/opt/microsoft/omsagent/tst/modules/syslog1/check_conf.py;                      source/code/troubleshooter/modules/syslog1/check_conf.py;                       644; root; root
```
Minor/Nitpick: I think you can make it "TstSyslog" or similar, "syslog1" sounds like a little bit like a temporary hack rather than finalized shipped product folder :)

aswatt replied:
Renamed it to 'syslog_tst':
https://github.com/microsoft/OMS-Agent-for-Linux/blob/09eb45de0d181e278fb5b1f83588ae0a2cc34324/installer/datafiles/base_omsagent.data#L176-L179

Dmitry-Matveev commented on source/code/troubleshooter/modules/connect/check_endpts.py:
```diff
@@ -0,0 +1,83 @@
+
+# check general internet connectivity
+def check_internet_connect():
+    if (check_endpt("bing.com") and check_endpt("google.com")):
```
Is there some generic network connection check we can use that is supposed to be more-less universal across linux systems, e.g. some update service URL, world clock or similar? I feel like bing.com/google.com can be blocked on purpose while internet in general is still available, but I do not know if there is better thing to use...

aswatt replied:
I haven't been able to find one is the issue, since technically customers only need to whitelist the endpoints given [on the Azure Docs page](https://docs.microsoft.com/en-us/azure/azure-monitor/platform/log-analytics-agent#firewall-requirements)

Dmitry-Matveev commented on source/code/troubleshooter/modules/errors_tsg.py:
```diff
@@ -0,0 +1,173 @@
+
+
+# get error codes from Troubleshooting.md
+def get_error_codes(err_type):
```
This method seems to impose hidden format dependency on Troubleshooting.md, therefore, changes in Troubleshooting.md formatting would cause this tool to fail. We may need to explicitly call this out in Troubleshooting.md file, so that any other developer touching it is aware of format requirements.

I see we already have baked-in codes for extension above, we may want to add more well-known baked-in codes to cover the case where GitHub Url is not available. Non-blocking, maybe, in the next iteration.

aswatt replied:
The main reason I wanted to grab them from the Troubleshooting.md file (rather than hard-coding them in) was it would automatically update with the addition of new error codes, but now with the addition of issues involving a firewall I'm not sure if hard-coding would just be better.

Right now there are 3 options regarding how to get the error codes from Troubleshooting.md:

1. Grab them from the link on Github (can fail with strict firewalls, relies on consistent formatting)
2. Grab the file upon bundle installation, like with the datafiles (will fail if OMS installation fails, relies on consistent formatting)
3. Hard-code the error codes and results in (will have to be updated manually if new error codes are added)

Which do you think is best? I'm genuinely unsure at this point

hestolz replied:
why not a combination? if you can't download them, fall back on the hardcoded file
i think i commented it in the past as well, but to avoid having to maintain changes in two files (the "real" Troubleshooting.md and the hardcoded one that goes in the TST), I assume you have some process for how to create a new `omsagent_tst.tar.gz` (if not you should). And in that process of targz creation, you can add the path like `docs/Troubleshooting.md`. That way any changes we make to the official page would end up in the next release.

The one caveat is you would still need to remember to regen a tar.gz if that .md changes, but you could make this part of the build process e.g. some step in Makefile is to create the tar.gz

aswatt replied:
Yeah, I currently have that for the log collector code, but it makes sense to have that for `docs/Troubleshooting.md` as well.
And actually regarding the `omsagent_tst.tar.gz`, that's an interesting thing that I'm not sure how to deal with. I need to create a new tar.gz file every time the log_collector or Troubleshooting.md files get updated, but installing via shell bundle doesn't even use `omsagent_tst.tar.gz` (it just grabs the files from their respective locations upon installation), so at this point the `omsagent_tst.tar.gz` file is only used when someone `wget`s it from Github. However, I don't really have a reliable way of keeping this up to date with any changes in the log collector or Troubleshooting.md, barring just watching for changes and manually updating them.
If I add the `tar` command to create the tar.gz file to the Makefile, it won't be updated on Github, which is the main issue. But I honestly can't think of a better way than just recreating the `omsagent_tst.tar.gz` file upon each release of a new version of the agent. What do you guys think?

aswatt commented:
I changed around how the troubleshooter is installed in bundle_skel.sh, here's an overview of how it works now:

- First, if the user runs with `--install` or `--upgrade` flags, then a variable `install_tst` is initialized (right at the beginning of the case):
https://github.com/microsoft/OMS-Agent-for-Linux/blob/d2625680b431576dc44d755eec545045f7983b75/installer/bundle/bundle_skel.sh#L1334-L1335

- After that, the script runs like normal, until `cleanup_and_exit()` is called (whether it has an error or not)

- At the start of `cleanup_and_exit()`, if the `install_tst` variable is initialized, it will go to install the troubleshooter:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/d2625680b431576dc44d755eec545045f7983b75/installer/bundle/bundle_skel.sh#L124-L134

- If the troubleshooter is already installed, it just skips the `wget` installation and prints out how to use it

- If not (most likely because OMS installation errored out in some way), it will attempt to install the troubleshooter via the `wget` installation:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/d2625680b431576dc44d755eec545045f7983b75/installer/bundle/bundle_skel.sh#L404-L409
And then print out how to use it upon successful install:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/d2625680b431576dc44d755eec545045f7983b75/installer/bundle/bundle_skel.sh#L462-L469

This should have no effect on OMS installation normally, since it doesn't happen until the end, but just in case I bookended the `install_troubleshooter` call with `set +e` and `set -e` so as not to raise errors, and no matter whether it errored or not it will return 0 (after printing out to console what went wrong).

Let me know what you think, or if this needs any tweaking! I was unsure about putting the `install_troubleshooter` call at the beginning or the end of `cleanup_and_exit()`, but I went for the beginning because I thought it would be better to have before the shell bundle exit statement.

hestolz commented on source/code/troubleshooter/modules/main.py:
```diff
@@ -7,7 +7,7 @@
 from heartbeat.heartbeat       import check_heartbeat
 from high_cpu_mem.high_cpu_mem import check_high_cpu_memory
-from syslog1.syslog            import check_syslog
+from syslog_tst.syslog            import check_syslog
```
nit indent

aswatt replied:
Sorry forgot to fix the indentation I had set up after changing `syslog1` to `syslog_tst`, fixed it here:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/1fcd4dd0d1cef4e18b71d295a048a3e4b7412dc9/source/code/troubleshooter/modules/main.py#L10
and here:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/1fcd4dd0d1cef4e18b71d295a048a3e4b7412dc9/installer/datafiles/base_omsagent.data#L176-L179

hestolz commented on installer/bundle/bundle_skel.sh:
```diff
@@ -1457,14 +1462,16 @@ case "$installMode" in
         OMI_EXIT_STATUS=$?
-	${OMI_SERVICE} reload
-	temp_status=$?
+	    ${OMI_SERVICE} reload
```
nit this section seems all weird indent

aswatt replied:
It looks like there were some tabs used instead of spaces? Sorry I thought I was fixing the weird indentation on my end. Changed all the tabs in that area to spaces for consistency:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/1fcd4dd0d1cef4e18b71d295a048a3e4b7412dc9/installer/bundle/bundle_skel.sh#L1469-L1479

hestolz commented on installer/bundle/bundle_skel.sh:
```diff
@@ -1328,6 +1331,8 @@ case "$installMode" in
         ;;
 
     I)
+        install_tst="yes" # set variable to install tst upon exit
```
nit by convention var name is caps

aswatt replied:
Sorry! Forgot about that, fixed here:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/1fcd4dd0d1cef4e18b71d295a048a3e4b7412dc9/installer/bundle/bundle_skel.sh#L1336

hestolz commented on installer/bundle/bundle_skel.sh:
```diff
@@ -412,54 +402,69 @@ install_if_program_does_not_exist_on_system()
+
+        # grab tst bundle
+        echo "Grabbing troubleshooter bundle from Github..."
+        wget $TST_PKG > /dev/null 2>&1
```
Kind of confused; with the latest changes, the only way that TST gets installed is via wget? Was this always the case? From what I remember, originally it was just bundled with the shell bundle, no wget involved. I think the general pattern for TST install or getting dynamic content like the .md can be try to download, but there should be a local backup.

aswatt replied:
Sorry, I need to make it more clear in the code, but the main way is actually getting installed with the shell bundle (all the files are in `base_omsagent.data`). I tested it, and on a successful install of OMS, the TST gets installed via shell bundle, which is why I have the check to see if the paths exist just a bit above. The only time the machine should be trying to install via `wget` is if OMS installation failed in such a way that the TST didn't get copied over, in which case having the TST on the machine would hopefully help (which is why I put it in `cleanup_and_exit()`).

aswatt replied:
Added clarification in the output + the comments in the code:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/1fcd4dd0d1cef4e18b71d295a048a3e4b7412dc9/installer/bundle/bundle_skel.sh#L404-L410

hestolz commented on source/code/troubleshooter/modules/errors_tsg.py:
```diff
@@ -25,19 +26,12 @@
         section = None
-        for line in ts_doc.readlines():
+        for line in ts_doc:
             line = line.decode('utf8').rstrip('\n')
```
does this work both py2/3?

aswatt replied:
Sorry yes you're right, I accidentally left that in after I changed Troubleshooting.md from just a urlopen check to checking both with the file and urlopen. I changed it so it only tries to decode if it's opening the file via url, which works on both Python2 and Python3 in both file and url form:
https://github.com/microsoft/OMS-Agent-for-Linux/blob/2da374a1507036bbd4a730dbd4f7f9ced60843be/source/code/troubleshooter/modules/errors_tsg.py#L35-L37
